[1단계 - 자동차 경주 구현] 상돌(이상진) 미션 제출합니다.

pricelees opened the pull request:
안녕하세요 제이미! 저는 6기 백엔드 코스에 참여중인 상돌(이상진)이라고 합니다!
(우선 늦긴 했지만, 새해 복 많이 받으세요(__))

미션은 첫 페어인 @shin-jisong 와 진행했고, 당연하게도 모든 의견이 일치되진 않았지만 서로 수용이 가능한 부분은 수용하며 즐거운 분위기 속에서 진행할 수 있었고, 페어프로그래밍의 즐거움을 알 수 있었던 것 같습니다😀

프리코스때 다른 참여자분들과 코드 리뷰를 꽤나 했었는데, 아무리 좋은 코드라고 해도 다른 사람이 작성한 코드를 읽는 것은 너무 어렵더라구요😥 읽어주셔서 너무 감사드리고, 제가  얘기하는 것을 너무 좋아하는 사람이라.. 어떠한 의견이던 즐겁게 받아들일 수 있으니 정말 편하게 봐주셨으면 좋겠습니다!

감사합니다.
이상진 드림



### 전반적인 구현 방향
1. 프리코스에서 MVC(라고 말하고 흉내만 낸..)를 사용해본 경험이 있으나, 이번 단계에선 MVC를 사용하지 않았습니다! 이번 미션은 기능 구현에 중점을 두었고, `클래스에 적절한 역할을 부여하는 것`을 중점으로 진행했습니다. 따라서 이번 미션은 별도로 패키지를 나누는 작업을 하진 않았습니다.

2. `getter`와 같은 단순한 기능을 제외한, 어느정도의 기능이 담긴 모든 메서드를 테스트하는 방향으로 진행했습니다.

3. 가독성을 중요하게 생각하고, 메서드 배치와 네이밍등에 대해 많이 고민하였습니다. 그래서 상수화가 안 되어있는 부분들도 있을텐데, 이 부분은 `상수로 보는 것 보다 그대로 읽는게 더 좋을 수도 있겠다`라는 판단 하에 진행한 부분입니다.

### 질문드리고 싶은 점
**1.어디까지 생각해야 할까요?**
페어를 잘 만나서인지, 코드를 작성할 때는 크게 의견이 갈리는 부분은 없었습니다. 거의 유일하게 의견이 갈렸던 부분이 기능 요구사항을 해석하는 부분이었습니다. 저는 `기능 요구사항에 명확히 작성된 부분이 없더라도, 어느정도 해석해서 진행하는게 맞다`라는 의견이었고, 페어는 `요구사항의 조건을 만족시키는 것이 중요하다`라는 관점이었습니다.

그래서, 예를 들면 요구사항에 작성된 `차량 이름의 길이는 5자 이하`를 저는 `그래도 차량 이름이니 적어도 1글자 이상은 들어가는게 맞다`로 해석했고, 페어는 `5글자 이하라고만 작성되었으니 0글자가 들어가도 작동되어야 한다`로 해석했는데, 이 부분에 대한 의견이 궁금합니다.

**2. 주석 처리는 어떠한 관점에서 하는게 좋을까요?**
프리코스때는 가급적 주석을 달지 않고 명확하게 해보자는 생각이었는데, 최근에는 조금이라도 모호한 부분이 있으면 주석을 작성하려고 하고 있습니다. 그래서 이번 단계에선 유틸성을 가지는 `Randoms` 클래스에 주석을 작성하였고, 주석 처리 기준에 대한 리뷰어님의 의견을 듣고 싶습니다.

**3. 예외 처리에 대한 관점을 들어보고 싶어요**
저는 콘솔에 입력되는 단계에서의 검증은 `정규 표현식`을 이용하여 검증하고, 도메인 객체를 생성할 때 한번 더 검증하도록 하고 있습니다. 

그런데 이번 미션을 하면서 문득 `CarName 클래스는 차량 이름을 명시하는데, 이 클래스의 생성자에 1~5글자라는 제약 조건을 넣어야 할까?` 라는 생각이 들었어요. `CarName`은 자동차 이름을 담는 클래스인데, 이 프로그램 상에서의 제약조건은 1~5글자라고 하더라도 실제 자동차 이름의 제약은 없기 때문에 `CarName`에서 이름 길이에 대한 유효성을 검증하는 것이 크게 의미가 없다는 생각이 들었습니다. 

실제로 입력을 받아들이는 `RacingGame` 클래스의 입력 단계에서 길이에 대한 검증을 하기 때문에, `CarName` 객체가 생성되는 시점에서의 차량 이름은 모두 유효한 값이 들어가기에 `CarName`에서의 유효성 검증은 테스트에만 쓰이는 상태입니다.

그래서 결론적으로는 리뷰어님의 예외 처리에 대한 관점이 너무 궁금합니다! 어떠한 의견이든 주시면 너무 감사할 것 같습니다!

shin-jisong committed:
docs: 기능 개요 정리

shin-jisong committed:
feat: 자동차 이름 정보를 저장하는 CarName 클래스

shin-jisong committed:
feat: 자동차 위치 정보를 저장하는 CarLocation 클래스

shin-jisong committed:
feat: 이름과 위치를 가지는 Car class

shin-jisong committed:
refactor: CarLocation 기능 수정

- randomNumber 체크를 Car class에서 진행하는 방향으로 변경

shin-jisong committed:
feat: carList를 가진 Cars Class

shin-jisong committed:
feat: 랜덤 숫자들을 만드는 Randoms class

shin-jisong committed:
feat: Cars Class 기능 구현 추가

- 모든 차를 전진시키는 함수
- 가장 많이 전진한 위치를 찾는 함수

shin-jisong committed:
feat: 경주를 주관하는 RacingGame class

shin-jisong committed:
feat: 자동차 경주 게임 실행 로직

shin-jisong committed:
docs: Race Class의 이름을 RacingGame으로 변경

shin-jisong committed:
docs: README.md 파일 수정

- 기능 요구 사항 수정
- 추가 정의 조건 추가
- 기능 구현 사항 설명 추가
- 예외 처리 사항 작성
- 콘솔 출력 예시 추가

shin-jisong committed:
refactor: 테스트를 위한 Car class에 location 파라미터 추가

shin-jisong committed:
refactor: 테스트를 위한 CarLocation class에 location 필드 및 해당 값 검증 추가

shin-jisong committed:
refactor: 팩토리 메서드 추가 및 상수화

- 이름만 파라미터로 사용하여 위치를 0으로 지정하여 Car 객체를 생성하는 팩토리 메서드 추가

shin-jisong committed:
feat: CarName 유효성 검증 추가

shin-jisong committed:
feat: RacingGame 유효성 검증 추가

shin-jisong committed:
comment: 주석 추가

shin-jisong committed:
test: CarLocation의 테스트 class

shin-jisong committed:
test: CarName의 테스트 class

shin-jisong committed:
test: Cars의 테스트 class

shin-jisong committed:
test: Car의 테스트 class

shin-jisong committed:
test: RacingGame의 테스트 class

shin-jisong committed:
style: 에러 출력 메시지 수정

shin-jisong committed:
sytle: 가독성을 위한 코드 순서 변경

shin-jisong committed:
sytle: 직관적인 의미 전달을 위한 함수 이름 변경

jamie9504 commented on README.md:
```diff
@@ -1,7 +1,106 @@
 자동차 경주 미션 저장소
 
-## 우아한테크코스 코드리뷰
+## 기능 요구사항 
```
`어디까지 생각해야 할까요?`에 대한 대답은 여기서 드릴 수 있을 것 같아요.

정답은 따로 없다고 생각하는데, 일단 요구사항은 정확히 구현해야 할 것이고 이 뒤로는 코치님들이나 페어와 맞춰보면 되겠죠?
코치님들이 알아서 하라고 하셨다면, 지금처럼 페어와 의견을 맞춰가면서 하면 되는거죠 😄 

상돌의 입장에서 개발하면 사용자에게 조금 더 편리함을 줄 수 있고,
상돌의 페어 입장에서 개발하면 정확한 요구사항을 구현할 수 있으니

무엇이 옳다 그르다보단 이렇게도 해보고 저렇게도 해보면 좋을 것 같네요 !
상돌 개인의 취향도 파악해보고, 취향이 맞지 않는 페어와 어떻게 의견을 조율해나가면 좋을지도 고민해보아요!
설득을 하는 것도 방법이고 설득을 당하는 것도 방법이랍니다.

pricelees replied:
좋은 의견 감사합니다 ㅎㅎ

jamie9504 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,40 @@
+public class Car {
+
+    private static final int MIN_NUMBER_FOR_MOVE = 4;
+    private final CarName carName;
+    private final CarLocation carLocation;
+
+
```
상수와 변수 사이에는 개행이 있으면 구분되어 보기가 좋아요 😄 

2칸의 개행은 의미가 있는 개행일까요? 별 의미가 없다면 2칸 이상의 개행을 지양해주세요!

pricelees replied:
하하.. 사실 어떤 의미가 아니라.. 실수입니다ㅜㅜ

jamie9504 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public static Car from(String name, int location) {
+        return new Car(
+                new CarName(name),
+                new CarLocation(location)
+        );
+    }
```
생성자를 호출할 때 개행을 하고 있는 부분들이 중간중간 보이는데요!
이 부분에 대해서도 개행을 하는 기준이 궁금하네요.

pricelees replied:
> 생성자를 호출할 때 개행을 하고 있는 부분들이 중간중간 보이는데요! 이 부분에 대해서도 개행을 하는 기준이 궁금하네요.

이 부분은 제가 작성한 부분인데요! 만약 `new Car(new CarName(name), new CarLocation(location));` 과 같이 작성했을때의 가독성이 그렇게 좋지 않다고 생각했어요. 이 경우는 `CarName`과 `CarLocation`의 이름이 짧아 비교적 괜찮다고는 생각하지만, 클래스 이름이 길어지는 경우 가독성이 너무 떨어지더라구요!

jamie9504 replied:
클래스 이름이 길어지는 경우 가독성이 떨어진다고 해주셨는데,
엔터를 쳐서 구분하는 것과 변수로 분리하는 것의 가독성을 비교해보는 것도 좋을 것 같아요~ 😄

pricelees replied:
> 클래스 이름이 길어지는 경우 가독성이 떨어진다고 해주셨는데, 엔터를 쳐서 구분하는 것과 변수로 분리하는 것의 가독성을 비교해보는 것도 좋을 것 같아요~ 😄

일단 진행하며 결론을 내린 것은, 한 줄로 써도 괜찮을 땐 한 줄에 넣자는 것입니다!!! ㅎㅎ

jamie9504 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,40 @@
+
+    public void move(int randomNumber) {
+        if (randomNumber >= MIN_NUMBER_FOR_MOVE) {
+            carLocation.increaseLocation();
```
carLocation.increaseLocation()은 뭔가 location이 두 번 중복되는 느낌이죠?

pricelees replied:
> carLocation.increaseLocation()은 뭔가 location이 두 번 중복되는 느낌이죠?

너무 맞습니다..ㅎㅎ 빠르게 반영하겠습니다

jamie9504 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,33 @@
+            int randomNumber = randomNumbers.get(index);
+            car.move(randomNumber);
+        }
+    }
```
이렇게 되면 Cars 객체만 보았을 때는, randomNumbers가 carLists 객체보다 사이즈가 크거나 같아야만 예외가 발생하지 않겠네요.
만약 상돌이 아닌 다른 개발자가 이 객체를 가져다 쓴다면, moveAll 메서드를 호출만 한다고 했을 때 이를 짐작할 수 있을까요?

pricelees replied:
> 이렇게 되면 Cars 객체만 보았을 때는, randomNumbers가 carLists 객체보다 사이즈가 크거나 같아야만 예외가 발생하지 않겠네요. 만약 상돌이 아닌 다른 개발자가 이 객체를 가져다 쓴다면, moveAll 메서드를 호출만 한다고 했을 때 이를 짐작할 수 있을까요?


**처음에 이 코드를 작성했을 때**는, `moveAll()`을 '유일'하게 호출하는 곳에서 `차량 리스트의 크기`라는 값을 불러와서 쓰기에 큰 문제가 없다고 생각했어요.

근데 이 커멘트를 보고 페어랑 논의를 해보니, `Cars` 클래스를 사용하는 새로운 클래스가 생긴다고 가정하면 문제가 생길 여지가 있다고 느꼈고, 이 부분에 대한 검증을 추가하는 것이 좋다는 생각으로 바뀌었습니다! 감사합니다 ㅎㅎ

jamie9504 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public List<Car> getCarList() {
+        return Collections.unmodifiableList(carList);
```
unmodifiableList를 사용하셨네요!
상돌이 unmodifiableList를 사용한 의도는 무엇인가요?
이를 생성자에서 사용하는 것과 getter에서 사용하는 것은 어떤 차이가 있을까요?

pricelees replied:
> unmodifiableList를 사용하셨네요! 상돌이 unmodifiableList를 사용한 의도는 무엇인가요? 이를 생성자에서 사용하는 것과 getter에서 사용하는 것은 어떤 차이가 있을까요?

`getCarList()`가 호출되는 곳에서의 사용에 대해 고민했어요. 이 메서드는 `현재의 경주 상태를 출력`하는 메서드와 `우승자를 출력`하는 메서드에서 호출이 되는데, 이 부분에서 `getCarList()`가 불변이 아니라면 이 게임에서 **가장 중요한 과정**인 `경주 상황과 우승자 출력`에 직접적으로 영향을 줄 수 있다고 생각했습니다!

jamie9504 replied:
unmodifiableList를 사용하면 carList는 불변이 될까요?
상돌이 생각하는 불변은 무엇인가요!

pricelees replied:
> unmodifiableList를 사용하면 carList는 불변이 될까요? 상돌이 생각하는 불변은 무엇인가요!

처음 생각은, `unmodifiableList`로 반환하면 getter를 통해 얻은 리스트를 수정하지 못할 것이라는 생각이었는데, 제이미의 말씀을 듣고  https://colevelup.tistory.com/47 를 찾아 보니 충분히 문제가 발생할 수 있다는 생각이 드네요!!ㅎㅎ 결론적으로는 2단계에선 이 getter를 삭제했습니다!

jamie9504 commented on src/main/java/RacingGame.java:
```diff
@@ -0,0 +1,113 @@
+import java.util.stream.Collectors;
+
+public class RacingGame {
+    private static final int MAX_RANDOM_NUMBER_RANGE = 9;
+    private static final String ERROR_MESSAGE = "[ERROR] 입력 형식이 올바르지 않습니다.";
```
위의 다른 클래스에선 문자열이 문자열 그대로 선언될 때도 있었는데, 여기서는 상수로 선언되어있네요!
상수로 선언하고 선언하지 않고를 결정한 기준은 무엇인가요?

pricelees replied:
> 위의 다른 클래스에선 문자열이 문자열 그대로 선언될 때도 있었는데, 여기서는 상수로 선언되어있네요! 상수로 선언하고 선언하지 않고를 결정한 기준은 무엇인가요?

**상수를 선언하기로 한 기준**은 다음과 같습니다!
1. 동일 클래스에서 **재사용이 되는 문자열**은 상수로 선언한다.
2. 재사용이 되지 않더라도, **그 값을 보고 의미를 바로 해석하기 힘들다면** 상수로 선언한다.

여기서 1번에 해당되는게 예외 메시지였고, 2번에 해당되는 것이 9라는 숫자였습니다!!

jamie9504 replied:
상돌의 기준을 알려줘서 고마워요.
그렇다면 예외 메시지의 경우, ERROR_MESSAGE라는 이름으로 어떤 내용인지 추측이 가능할까요?

pricelees replied:
> 상돌의 기준을 알려줘서 고마워요. 그렇다면 예외 메시지의 경우, ERROR_MESSAGE라는 이름으로 어떤 내용인지 추측이 가능할까요?

콘솔에 출력되는 예외메시지는 모두 동일해서 통일했는데, 가독성 이슈가 있는 것 같습니다..ㅋㅋㅋ ㅜㅜ

jamie9504 commented on src/main/java/RacingGame.java:
```diff
@@ -0,0 +1,113 @@
+        if (hasInvalidNameLength(names.split(","))) {
+            throw new IllegalArgumentException(ERROR_MESSAGE);
+        }
+    }
```
validate를 Game에서 해주었을 때 어떤 장/단점이 있을 수 있을지 고민해보면 좋을 것 같아요!
validate를 하는 주체가 누가 되는 것이 좋을지는 항상 나오는 단골 주제이죠!

NAMES_REGEX는 조금 더 명확하게 이름을 가져가봐도 좋을 것 같네요 (NAMES_VALID_REGEX 라던가...)
요건 취향 차이일 수도 있어요 😃

pricelees replied:
> validate를 Game에서 해주었을 때 어떤 장/단점이 있을 수 있을지 고민해보면 좋을 것 같아요! validate를 하는 주체가 누가 되는 것이 좋을지는 항상 나오는 단골 주제이죠!
> 
> NAMES_REGEX는 조금 더 명확하게 이름을 가져가봐도 좋을 것 같네요 (NAMES_VALID_REGEX 라던가...) 요건 취향 차이일 수도 있어요 😃

**이름**
이름은 더욱 더 신경써보겠습니다 ㅎㅎ 다시보니 너무 애매하네요..ㅜ

**Validate**
Validate를 Game에서 진행했던 것은, 두 가지 이유였어요.
1. MVC 구조를 사용했다면 입력을 받는 클래스를 만들고 진행했을 것 같지만, MVC를 사용하지 않았고 그래서 추가적으로 validation을 하는 클래스까지 만들고 싶지 않았어요.
2. 이 검증은 이 게임에서만 사용된다고 생각했기에 `RacingGame`에서 진행하는 것이 가장 흐름상 맞다는 생각이었어요. 

질문 드렸던 `CarName`에서의 예외 처리와 비교해서 생각해보면, 

`CarName`은 다른 `Car` 와 관련된 게임이 있다면 충분히 재사용할수 있다고 생각했어요. 그래서 `CarName에서의 예외 처리를 하는 것이 맞는가?` 에 대한 생각을 했었어요.

그런데 `RacingGame`은 재사용되지 않을 것이라고 생각했어요. 만약에 `검증 조건이 바뀐다고 생각했을 때, 검증 관련 코드를 어디서 찾아야 할까?` 라고 생각해보니 `RacingGame`에 검증을 넣는 것이 이 상황에선 가장 직관적으로 느껴졌었던 것 같습니다!

jamie9504 replied:
> MVC 구조를 사용했다면 입력을 받는 클래스를 만들고 진행했을 것 같지만, MVC를 사용하지 않았고 그래서 추가적으로 validation을 하는 클래스까지 만들고 싶지 않았어요.

저도 꼭 MVC 구조를 사용할 필요는 없다고 생각해요.

현재 상돌의 구조에서 RacingGame을 콘솔이 아닌 다른 것으로 입력값을 받아온다고 할 때, (ex. 저장된 text 값이라던가)
수정의 범위가 어디까지일지 고민해보는 것도 좋을 것 같아요 :) 

> 이 검증은 이 게임에서만 사용된다고 생각했기에 RacingGame에서 진행하는 것이 가장 흐름상 맞다는 생각이었어요.
👍

jamie9504 commented on src/main/java/RacingGame.java:
```diff
@@ -0,0 +1,113 @@
+        System.out.println("실행 결과");
+    }
+
+    protected List<Integer> createRandomNumber(int carCount) {
```
protected ! 를 쓰면 어떤 점이 좋을까요?

pricelees replied:
> protected ! 를 쓰면 어떤 점이 좋을까요?

이 메서드를 protected로 선언한 이유는, 사실 `테스트`가 가장 큰 목적이었고, `RacingGameTest`에서 **익명 클래스**를 이용하여 `createRandomNumber()`가 반환하는 값을 임의로 설정하는데 사용되었어요. 

그런데 오늘 **피드백**강의를 듣고나니, 이번 미션의 목표가 테스트다보니 **너무 테스트에만 집중했다**라는 생각이 많이 들게 되더라구요..ㅠ 테스트를 위해 프로덕션을 수정하는게 맞는지에 대해 많이 고민한 것 같아요.

jamie9504 replied:
테스트를 위해 프로덕션을 수정하는 것이 맞는지도 취향이 듬뿍 담기는 부분이죠.
많이 고민하셨으니 상돌만의 기준이 생겼으면 좋겠네요!

jamie9504 commented on src/main/java/RacingGame.java:
```diff
@@ -0,0 +1,113 @@
+        System.out.println();
+    }
+
+    private void announceWinners(Cars cars) {
```
메서드명은 일반적으로 자주 쓰는 동사를 사용해주는 것이 조금 더 직관적으로 보일 수도 있어요.
물론 print announce 모두 좋은 의미가 될 수 있다고 생각해요.

printWinners와 announceWinners는 각각 상돌에게 어떤 느낌을 주는지 궁금하네요 😄

pricelees replied:
> 메서드명은 일반적으로 자주 쓰는 동사를 사용해주는 것이 조금 더 직관적으로 보일 수도 있어요. 물론 print announce 모두 좋은 의미가 될 수 있다고 생각해요.
> 
> printWinners와 announceWinners는 각각 상돌에게 어떤 느낌을 주는지 궁금하네요 😄

이 부분은 사실 페어와 `print`와 `announce`를 둘다 써놓고 결정한 부분인데, 그때의 생각은 `우승자`는 `출력`보다 `발표`라고 작성하는게 맥락에 맞다고 생각했어요. 

그런데 이걸 보고 다시 보니 다른 출력과 관련된 메서드명은 모두 print로 작성했다는 것을 알게 되었고, 전체 맥락에선 print가 더 직관적으로 느껴지겠다는 생각을 하게 되었습니다!

jamie9504 reviewed (changes requested):
안녕하세요. 상돌 😄. 리뷰어 제이미입니다.

우테코 첫 PR 축하드리고, 상돌도 새해 복 많이 받으세요 🎉 

저도 상돌의 의견이 궁금해서 궁금한 부분 코멘트로 조금 남겨두었어요.
개인적으로 코드에는 정답은 없다고 생각하고 많은 사람들과 의견을 나눠본 후 자신만의 의견을 구축해보는 것이 좋다!고 생각하기 때문에, 페어라던가 같은 조 사람이라던가 다른 크루들이랑 많이 이야기를 나눠보면 좋을 것 같아요.

질문주신 부분은 아래에 답변 남겨둘께요.

---

1. 코드쪽 코멘트에 관련 부분 달아두었어요.

2. `조금이라도 모호한 부분이 있으면 주석을 작성` 이라고 해주셨는데, 왜 조금이라도 모호했을지 고민해보는 것도 좋을 것 같아요.
과연 상돌이 명명한게 최선의 최적의 네이밍이었을지, 코드로는 다 설명이 불가능했을지 고민해볼까요?

3. `CarName 클래스는 차량 이름을 명시하는데, 이 클래스의 생성자에 1~5글자라는 제약 조건을 넣어야 할까?` 이 부분은 상돌이 만든 CarName이라는 친구의 역할에 대해 고민해보면 좋을 것 같네요.
상돌이 만든 프로그램 세계에 살고 있는 CarName이라는 친구가 가지는 역할에 이름 제한이 들어가는 것이 맞을지요. 프로그램 밖의 CarName과 역할이 같을 필요는 없으니까요 😄 

위 부분을 고민해보고 적어주시면 다음 요청 때 추가로 답변드리도록 할께요 😄

Commit:
style(Car) : 메서드 의미 중복 제거

< description >
- carLocation.increaseLocation()은 의미가 겹친다고 생각했고, carLocation.increase()로 해당 부분 수정
- getter에서도 역시 의미가 겹친다고 생각해서, 이 부분에 대한 고민이 필요할 것 같음

Commit:
Revert "style(Car) : 메서드 의미 중복 제거"

This reverts commit a2af457c73fe682349dbe8a606f1444d0467dbe3.

Commit:
style(CarLocation) : 메서드 의미 중복 제거

Commit:
refactor(Cars) : 모든 차량을 움직이는 moveAll()에 대한 유효성 검증 추가

Commit:
test(CarsTest) : Cars의 유효성 검증 추가에 따른 테스트 추가

Commit:
style(RacingGame) : 출력과 관련된 메서드명의 접두사를 print로 수정

pricelees reviewed (commented):
안녕하세요 제이미! 즐거운 금요일입니다 ㅎㅎ

말씀해주신 부분들과 오늘 피드백 강의까지 해서 많은 생각을 할 수 있었어요..ㅎㅎ 
일단 질문주신 부분은 모두 답변드렸어요! 추가적으로 답변이 필요한 부분이 있다면 다음번에 바로 반영하도록 해보겠습니다! 

감사합니다.

Commit:
style(Car) : 메서드 의미 중복 제거

< description >
- increaseLocation()의 Location의 중복 제거

jamie9504 commented on README.md:
```diff
@@ -1,7 +1,106 @@
 
 자동차 경주 미션 저장소
 
-## 우아한테크코스 코드리뷰
```
이번 추가된 커밋의 경우, 기존과 커밋 이름이 조금 달라졌는데요!
수정된 클래스명을 명시해주었네요!

클래스명을 명시하는 커밋과 명시하지 않는 커밋은 어떤 차이가 있을지 고민해보고
상돌은 어떤 것을 선택하는게 좋다고 생각하는지 알려주세요 :)

pricelees replied:
> 이번 추가된 커밋의 경우, 기존과 커밋 이름이 조금 달라졌는데요! 수정된 클래스명을 명시해주었네요!
> 
> 클래스명을 명시하는 커밋과 명시하지 않는 커밋은 어떤 차이가 있을지 고민해보고 상돌은 어떤 것을 선택하는게 좋다고 생각하는지 알려주세요 :)

사실 이게 궁금했습니다..ㅎㅎ 저는 두 방법 모두 해봤을 때 후자가 커밋 메시지를 조금 더 간결하게 작성하고, 어떤 클래스인지를 조금 더 직관적으로 알 수 있다고 생각합니다!

jamie9504 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,40 @@
+public class Cars {
+    private final List<Car> carList;
+
+    public Cars(List<Car> carList) {
+        this.carList = carList;
+    }
+
+    public void moveAll(List<Integer> randomNumbers) {
```
어떤 곳은 List일 때 -List를 어떤 곳은 -s를 붙여줬네요!
상돌은 변수명을 지어줄 때 어떤 기준으로 짓는지 궁금해요!
어떤 것이 옳다보다는 일관성이 있고 그 일관성을 설명할 수 있어야, 코드를 읽을 때 직관적이라고 생각합니다 😄 

또한 Cars의 입장에서, 숫자가 randomNumbers인지 알 필요가 있을지 고민해봐도 좋을 것 같아요.
random한 숫자인지를 파라미터명으로 보장을 해줘야 하는 걸까요?

pricelees replied:
> 어떤 곳은 List일 때 -List를 어떤 곳은 -s를 붙여줬네요! 상돌은 변수명을 지어줄 때 어떤 기준으로 짓는지 궁금해요! 어떤 것이 옳다보다는 일관성이 있고 그 일관성을 설명할 수 있어야, 코드를 읽을 때 직관적이라고 생각합니다 😄
> 
> 또한 Cars의 입장에서, 숫자가 randomNumbers인지 알 필요가 있을지 고민해봐도 좋을 것 같아요. random한 숫자인지를 파라미터명으로 보장을 해줘야 하는 걸까요?

의견 감사합니다!! 너무 요구사항대로만 생각했던 것 같습니다..ㅠ 다시 보니 Car의 이동 조건이 randomNumber에만 한정된다는 느낌이 들었습니다!!ㅎㅎ

jamie9504 commented on src/main/java/RacingGame.java:
```diff
@@ -0,0 +1,113 @@
+        String winners = cars.getCarList().stream()
+                .filter(car -> car.getCarLocation() == cars.getMaxLocation())
+                .map(Car::getCarName)
+                .collect(Collectors.joining(", ")) + "가 최종 우승했습니다.";
+        ;
```
이런 작은 부분도 신경써주세요 ! :)

(불필요한 개행, 불필요한 문자열 등)

pricelees replied:
ㅠㅠ 더 많이 신경쓰겠습니다 저도 못봤던 부분인데 정말 자세하게 봐주셔서 감사드려요!!

jamie9504 commented on src/main/java/Randoms.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+import java.util.Random;
+
+public class Randoms {
```
클래스명이 Randoms인데, Randoms라는 클래스는 어떤 일을 할 것으로 예측할 수 있을까요?
해당 이름을 지어주게 된 이유가 있을까요?

pricelees replied:
> 클래스명이 Randoms인데, Randoms라는 클래스는 어떤 일을 할 것으로 예측할 수 있을까요? 해당 이름을 지어주게 된 이유가 있을까요?

메서드명에 `number`가 들어가기에, RandomNumber와 같은 클래스명으로 하면 number가 중복되는 느낌이었습니다..ㅎㅎ 2단계때는 조금 더 반영해보았습니다!

jamie9504 commented on src/main/java/Randoms.java:
```diff
@@ -0,0 +1,29 @@
+public class Randoms {
+    private static final Random random = new Random();
+
+    /**
+     * @param count        뽑을 숫자 개수
+     * @param endInclusive 숫자 범위
+     * @return 0부터 endInclusive까지의 count만큼의 랜덤 숫자 리스트
+     */
```
메서드의 시그니처만 보았을 때 주석의 내용 중 짐작을 할 수 없는 것은 어떤 것이 있나요?
주석을 사용하지 않고 짐작할 수 있도록 시그니처를 바꿀 수 있을지와 주석으로 처리했을 때의 장/단점을 고민해볼까요!

pricelees replied:
> 메서드의 시그니처만 보았을 때 주석의 내용 중 짐작을 할 수 없는 것은 어떤 것이 있나요? 주석을 사용하지 않고 짐작할 수 있도록 시그니처를 바꿀 수 있을지와 주석으로 처리했을 때의 장/단점을 고민해볼까요!

자바 내부 함수를 사용하기도 하고, 코드 자체도 명확하다고 생각해서 주석을 처음에는 생략했었는데, `유틸 클래스`에는 모두 주석을 달아보자는 생각이었습니다 ㅎㅎ 제가 명확하다고 생각해도 꼭 그게 아니더라구요

jamie9504 reviewed (approved):
안녕하세요, 상돌 😄 리뷰어 제이미입니다!

이번에도 코멘트를 조금 남겨두었어요!

1단계는 이만 머지합니다! 2단계에서 함께 작업해주세요 😉

화이팅 💪

pricelees reviewed (commented):
제이미 안녕하세요!
말씀해주신 부분은 모두 꼼꼼히 생각해봤고, 2차때 반영하도록 많이 노력해봤습니다!
두 번이나 봐주실 줄은 몰랐는데..ㅠㅠ 꼼꼼히 봐주셔서 너무 감사드려요!😊
